artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/impl/Page.java

```diff
@@ -120,105 +110,121 @@ public void setLiveCache(LivePageCache pageCache) {
-         return read(storage, activeMQBuffer, messages);
-      } finally {
-         //unmap the file after read it to avoid GC to take care of it
-         PlatformDependent.freeDirectBuffer(mappedByteBuffer);
+   private ByteBuffer readIntoFileBufferIfNecessary(SequentialFile file,
+                                                    ByteBuffer fileBuffer,
+                                                    int requiredBytes) throws Exception {
+      final int remaining = fileBuffer.remaining();
+      if (remaining < requiredBytes) {
+         final int capacity = fileBuffer.capacity();
+         if (capacity >= requiredBytes) {
+            final int bytesToBeRead = requiredBytes - remaining;
+            final int availableCapacity = capacity - fileBuffer.limit();
+            if (availableCapacity >= bytesToBeRead) {
+               final int originalPosition = fileBuffer.position();
+               fileBuffer.limit(capacity);
+               file.read(fileBuffer);
+               fileBuffer.position(originalPosition);
+            } else {
+               compactAndReadIntoFileBuffer(file, fileBuffer);
+            }
+         } else {
+            fileBuffer = allocateAndReadIntoFileBuffer(file, fileBuffer, requiredBytes);
+         }
       }
+      return fileBuffer;
    }
 
-   private int read(StorageManager storage, ActiveMQBuffer fileBuffer, List<PagedMessage> messages) {
-      int readMessages = 0;
-      while (fileBuffer.readable()) {
-         final int position = fileBuffer.readerIndex();
-
-         byte byteRead = fileBuffer.readByte();
-
-         if (byteRead == Page.START_BYTE) {
-            if (fileBuffer.readerIndex() + DataConstants.SIZE_INT < fileBuffer.capacity()) {
-               int messageSize = fileBuffer.readInt();
-               int oldPos = fileBuffer.readerIndex();
-               if (fileBuffer.readerIndex() + messageSize < fileBuffer.capacity() && fileBuffer.getByte(oldPos + messageSize) == Page.END_BYTE) {
-                  PagedMessage msg = new PagedMessageImpl(storageManager);
-                  msg.decode(fileBuffer);
-                  byte b = fileBuffer.readByte();
-                  if (b != Page.END_BYTE) {
-                     // Sanity Check: This would only happen if there is a bug on decode or any internal code, as
-                     // this
-                     // constraint was already checked
-                     throw new IllegalStateException("Internal error, it wasn't possible to locate END_BYTE " + b);
-                  }
+   private List<PagedMessage> readFromSequentialFile(StorageManager storage) throws Exception {
+      final List<PagedMessage> messages = new ArrayList<>();
+      final int fileSize = (int) file.size();
+      file.position(0);
+      final int minimumMessagePersistentSize = DataConstants.SIZE_INT + 2;
+      int position = 0;
+      ByteBuffer fileBuffer = null;
+      try {
+         while ((fileSize - position) >= minimumMessagePersistentSize) {
```

Comment:
What was wrong with the previous reading logic, where we would just check for START_BYTE?

This might be right, but did you do enough tests on this to make sure this is ok?

I'm afraid of changing logic here on a years old tested code for no reason. If you found a reason for this I would like to know.

Reply:
But I'm checking START_BYTE as before, the logic on the single message hasnìt changed at all.
In the new logic I'm moving a window of ByteBuffer on the file regions in order to be able to reduce I/O and buffer copies/allocations as much as possible.